xLights/sequencer/tabSequencer.cpp

```diff
@@ -1502,24 +1502,75 @@ void xLightsFrame::UpdateEffect(wxCommandEvent& event)
                 mSequenceElements.get_undo_mgr().CaptureModifiedEffect(element->GetModelName(),
                                                                        el->GetIndex(),
                                                                        ef);
-                
+
                 el->GetEffect(j)->SetSettings(effectText, true);
                 el->GetEffect(j)->SetEffectIndex(effectIndex);
                 el->GetEffect(j)->SetEffectName(effectName);
                 el->GetEffect(j)->SetPalette(palette);
-                
+
+                startms = std::min(startms, el->GetEffect(j)->GetStartTimeMS());
+                endms = std::max(endms, el->GetEffect(j)->GetEndTimeMS());
+            }
+        }
+
+        if(startms <= endms) {
+            playType = PLAY_TYPE_EFFECT;
+            playStartMS = -1;
+            RenderEffectForModel(element->GetModelName(),startms,endms);
+        }
+    }
+    mainSequencer->PanelEffectGrid->ForceRefresh();
+}
+
+void xLightsFrame::RandomizeEffect(wxCommandEvent& event)
+{
+    mSequenceElements.get_undo_mgr().CreateUndoStep();
+
+    for(size_t i=0;i<mSequenceElements.GetVisibleRowInformationSize();i++) {
+        Element* element = mSequenceElements.GetVisibleRowInformation(i)->element;
+        EffectLayer* el = mSequenceElements.GetVisibleEffectLayer(i);
+
+        int startms = 99999999;
+        int endms = -1;
+
+        for(int j=0;j< el->GetEffectCount();j++) {
+            if(el->GetEffect(j)->GetSelected() != EFFECT_NOT_SELECTED)  {
+                std::string effectName = el->GetEffect(j)->GetEffectName();
+                int effectIndex = el->GetEffect(j)->GetEffectIndex();
+                std::string settings, palette;
+
+                settings = EffectsPanel1->GetRandomEffectString(effectIndex);
+                palette = colorPanel->GetRandomColorString();
```

Comment:
these 2 lines and the last line of this method are the only diff between the randomize and update function above.